Report error when there are duplicate definitions of a strong symbol

Error reporting is something that we've neglected a bit. At the moment, we don't report an error when there are multiple definitions of a symbol.

There are a few nuances to multiple-definition rules:

* It's only a problem if there's more than one strong definition. Multiple weak definitions, optionally with one strong definition is fine.
* It's only a problem if the object containing the symbol is loaded. e.g. it's fine to have strong definitions in archive entries that duplicate other symbols and so long as those archive entries aren't loaded, it's OK.

Selecting from multiple symbol definitions is handled in `resolve_alternative_symbol_definitions`, so that would be the place to start looking.

If there are multiple symbols that have duplicate definitions, it might be good to report all of them, although probably not essential.

We should have a test for this. See `undefined_symbols.c` for a similar kind of test to use as a reference.

Hi, @davidlattimore. First let me apologize for not addressing this issue until now. I've been busy the past few weeks.

Now, I made some changes to achieve better error reporting. And in an attempt to add some tests, I created the following test code with multiple strong symbols:

```c
//#AbstractConfig:default
//#Config:executable:default
//#SkipLinker:ld
//#CompArgs:-fcommon
//#LinkerDriver:none
//#LinkArgs:--shared --allow-multiple-definition
//#ExpectError:redefinition of

int test_func(void) {
    return 0;
}

int test_func(void) {
    return 1;
}

int main(void) {
    return test_func();
}
```

This is of course invalid code, so `#SkipLinker:ld` is added. This gives the following message on the test, so I think it is trying to link with wild: 
> Error: Failed to build program `multiple_strong_symbols.c` with linker `wild` config `executable`

However, it seems that wild is called via gcc, even though I added `#LinkerDriver:none`. This seems odd, am I doing something wrong in my configuration?
> Compilation failed:  gcc -Werror=attributes -c /home/lapla/repos/wild/wild/tests/sources/multiple_strong_symbols.c -fcommon -o /home/lapla/repos/wild/wild/tests/build/multiple_strong_symbols.executable-host-739f6959b3fa36d9.o

<details><summary>the output from integration_tests</summary>

```
     Running tests/integration_tests.rs (target/debug/deps/integration_tests-d675bcfb9842e0ea)

running 1 test
/home/lapla/repos/wild/wild/tests/sources/multiple_strong_symbols.c:13:5: error: redefinition of ‘test_func’
   13 | int test_func(void) {
      |     ^~~~~~~~~
/home/lapla/repos/wild/wild/tests/sources/multiple_strong_symbols.c:9:5: note: previous definition of ‘test_func’ with type ‘int(void)’
    9 | int test_func(void) {
      |     ^~~~~~~~~
Error: Failed to build program `multiple_strong_symbols.c` with linker `wild` config `executable`

Caused by:
    Compilation failed:  gcc -Werror=attributes -c /home/lapla/repos/wild/wild/tests/sources/multiple_strong_symbols.c -fcommon -o /home/lapla/repos/wild/wild/tests/build/multiple_strong_symbols.executable-host-739f6959b3fa36d9.o
test integration_test::program_name_1___multiple_strong_symbols_c__ ... FAILED

failures:

failures:
    integration_test::program_name_1___multiple_strong_symbols_c__
```
</details>

--- a/libwild/src/symbol_db.rs
+++ b/libwild/src/symbol_db.rs
@@ -28,6 +28,7 @@ use ahash::HashMap;
 use ahash::RandomState;
 use anyhow::Context;
 use anyhow::bail;
+use crossbeam_queue::SegQueue;
 use itertools::Itertools;
 use object::LittleEndian;
 use object::read::elf::Sym as _;
@@ -37,6 +38,7 @@ use rayon::iter::IntoParallelRefMutIterator;
 use rayon::iter::ParallelIterator;
 use std::collections::hash_map;
 use std::fmt::Display;
+use std::fmt::Write;
 use std::mem::replace;
 use std::mem::take;
 
@@ -596,25 +598,27 @@ pub(crate) fn resolve_alternative_symbol_definitions<'data>(
     resolved: &[ResolvedGroup],
 ) -> Result {
     let mut buckets = take(&mut symbol_db.buckets);
+    let error_queue = SegQueue::new();
 
     let replacements: Vec<Vec<(SymbolId, SymbolId)>> = buckets
         .par_iter_mut()
         .map(|bucket| {
-            let mut replacements = Vec::new();
-
-            let mut add_replacement = |a, b| {
-                replacements.push((a, b));
-            };
+            let mut local_replacements = Vec::new();
 
             for (first, alternatives) in replace(
                 &mut bucket.alternative_definitions,
                 HashMap::with_hasher(RandomState::new()),
             ) {
-                let selected = select_symbol(symbol_db, first, &alternatives, resolved);
-                add_replacement(first, selected);
-
-                for alt in alternatives {
-                    add_replacement(alt, selected);
+                match select_symbol(symbol_db, first, &alternatives, resolved) {
+                    Ok(selected) => {
+                        local_replacements.push((first, selected));
+                        for alt in alternatives {
+                            local_replacements.push((alt, selected));
+                        }
+                    }
+                    Err(err) => {
+                        error_queue.push(err);
+                    }
                 }
             }
 
@@ -622,17 +626,35 @@ pub(crate) fn resolve_alternative_symbol_definitions<'data>(
                 &mut bucket.alternative_versioned_definitions,
                 HashMap::with_hasher(RandomState::new()),
             ) {
-                let selected = select_symbol(symbol_db, first, &alternatives, resolved);
-                add_replacement(first, selected);
-
-                for alt in alternatives {
-                    add_replacement(alt, selected);
+                match select_symbol(symbol_db, first, &alternatives, resolved) {
+                    Ok(selected) => {
+                        local_replacements.push((first, selected));
+                        for alt in alternatives {
+                            local_replacements.push((alt, selected));
+                        }
+                    }
+                    Err(err) => {
+                        error_queue.push(err);
+                    }
                 }
             }
-
-            replacements
+            local_replacements
         })
         .collect();
+
+    let mut duplicate_errors: Vec<anyhow::Error> = error_queue.into_iter().collect();
+    duplicate_errors.sort_by_key(|e| e.to_string());
+
+    if !duplicate_errors.is_empty() {
+        let error_details = duplicate_errors
+            .iter()
+            .map(|e| e.to_string())
+            .collect::<Vec<_>>()
+            .join("\n");
+        return Err(anyhow::Error::msg(format!(
+            "Duplicate symbols detected: {error_details}"
+        )));
+    }
 
     for (a, b) in replacements.iter().flat_map(|r| r.iter()) {
         symbol_db.replace_definition(*a, *b);
@@ -650,23 +672,23 @@ fn select_symbol(
     symbol_id: SymbolId,
     alternatives: &[SymbolId],
     resolved: &[ResolvedGroup],
-) -> SymbolId {
-    let first_strength = symbol_db.symbol_strength(symbol_id, resolved);
-    if first_strength == SymbolStrength::Strong {
-        return symbol_id;
-    }
+) -> Result<SymbolId, anyhow::Error> {
     let mut max_common = None;
-    for &alt in alternatives.iter().rev() {
-        // Dynamic symbols, even strong ones, don't override non-dynamic weak symbols.
+    let mut strong_symbols = Vec::new();
+
+    let all_symbols = std::iter::once(symbol_id).chain(alternatives.iter().copied());
+
+    for alt in all_symbols {
         if symbol_db
             .symbol_value_flags(alt)
             .contains(ValueFlags::DYNAMIC)
         {
             continue;
         }
+
         let strength = symbol_db.symbol_strength(alt, resolved);
         match strength {
-            SymbolStrength::Strong => return alt,
+            SymbolStrength::Strong => strong_symbols.push(alt),
             SymbolStrength::Common(size) => {
                 if let Some((previous_size, _)) = max_common {
                     if size <= previous_size {
@@ -678,19 +700,46 @@ fn select_symbol(
             _ => {}
         }
     }
+
+    if strong_symbols.len() > 1 {
+        let already_defined_in = symbol_db.file_id_for_symbol(*strong_symbols.first().unwrap());
+
+        return Err(anyhow::Error::msg(format!(
+            "{}{}, previously defined in {}",
+            symbol_db.symbol_name_for_display(symbol_id),
+            strong_symbols
+                .iter()
+                .skip(1)
+                .fold(String::new(), |mut output, &s| {
+                    let fid = symbol_db.file_id_for_symbol(s);
+                    write!(output, ", defined in {}", symbol_db.file(fid)).unwrap();
+                    output
+                }),
+            symbol_db.file(already_defined_in)
+        )));
+    }
+
+    if let Some(&strong_symbol) = strong_symbols.first() {
+        return Ok(strong_symbol);
+    }
+
     if let Some((_, alt)) = max_common {
-        return alt;
+        return Ok(alt);
     }
+
+    let first_strength = symbol_db.symbol_strength(symbol_id, resolved);
     if first_strength != SymbolStrength::Undefined {
-        return symbol_id;
+        return Ok(symbol_id);
     }
+
     for &alt in alternatives.iter().rev() {
         let strength = symbol_db.symbol_strength(alt, resolved);
         if strength != SymbolStrength::Undefined {
-            return alt;
+            return Ok(alt);
         }
     }
-    symbol_id
+
+    Ok(symbol_id)
 }
 
 #[derive(PartialEq, Eq, Clone, Copy)]

--- a/wild/tests/integration_tests.rs
+++ b/wild/tests/integration_tests.rs
@@ -2158,7 +2158,8 @@ fn integration_test(
         "tls-local-exec.c",
         "undefined_symbols.c",
         "whole_archive.c",
-        "shared.c"
+        "shared.c",
+        "duplicate_strong_symbols.c"
     )]
     program_name: &'static str,
     #[allow(unused_variables)] setup_symlink: (),

--- a/wild/tests/sources/duplicate_strong_symbols.c
+++ b/wild/tests/sources/duplicate_strong_symbols.c
@@ -1,0 +1,13 @@
+//#Config:dup
+//#SkipLinker:ld
+//#Object:duplicate_strong_symbols2.c
+//#Object:exit.c
+//#ExpectError:Duplicate symbols
+
+int test_func(void) {
+    return 0;
+}
+
+void _start(void) {
+    test_func();
+}

--- a/wild/tests/sources/duplicate_strong_symbols2.c
+++ b/wild/tests/sources/duplicate_strong_symbols2.c
@@ -1,0 +1,3 @@
+int test_func(void) {
+    return 1;
+}

